docs/root/intro/arch_overview/matching/matching_api.rst

```diff
@@ -0,0 +1,112 @@
+.. _arch_overview_matching_api:
+
+Matching API
+============
+
+Envoy makes use of a matching API to allow the various subsystems to express actions that should be performed based on incoming data.
+
+The matching API is designed as a tree structure to allow for sublinear matching algorithms, and make heavy use of extension points to make it
+easy to extend to different inputs based on protocol or environment data, sublinear matchers and direct matchers. 
+
+Within supported environments (currently only HTTP filters), a wrapper proto can be used to instantiate a matching filter associated with the
+wrapped structure:
+
+.. code-block:: yaml
+
+    "@type": type.googleapis.com/envoy.extensions.common.matching.v3.ExtensionWithMatcher
+    extension_config:
+        name: response-filter-config
+        typed_config:
+            "@type": type.googleapis.com/test.integration.filters.SetResponseCodeFilterConfig
+            code: 403
+    matcher:
+        matcher_tree:
+            input:
+                name: request-headers
+                typed_config:
+                    "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
+                    header_name: some-header
+            exact_match_map:
+                # Note this additional indirection; this is a workaround for Protobuf oneof limitations.
+                map:
+                    skip_filter: # This is the header value we're trying to match against.
+                        action:
+                            name: skip
+                            typed_config:
+                                "@type": type.googleapis.com/envoy.extensions.filters.common.matcher.action.v3.SkipFilter
+
+The above example wraps a HTTP filter (the SetResponseCode filter) in an ExtensionWithMatcher, allowing us to define
```

Comment:
Nit: prefer some RST formatting for filer proper names, e.g. `*SetResponseCode*`.

Reply:
should this be a link to the relevant filter?